File: VariousArtists/string_permutations.rb
# permute string

class String
    def permute
        return [self] if self.size == 1
        permutations = self[1..-1].permute
        char = self[0]
        result = []
        permutations.each { |x|
            for i in 0..self[1..-1].size
                result << "#{x[0, i]}#{char}#{x[i..-1]}"
            end
        }

        result
    end
end

puts "gerula".permute.inspect
puts "abc".permute.inspect

File: VariousArtists/string_permutations.cs
// Given a string, collect all permutations
//

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

static class Program
{
    static List<String> Permute(this String s)
    {
        var result = new List<String>();
        Permutations(new StringBuilder(s), 0, result);
        return result;
    }

    static void Permutations(StringBuilder s, int start, List<String> results)
    {
        if (start == s.Length)
        {
            results.Add(s.ToString());
            return;
        }

        for (var i = start; i < s.Length; i++)
        {
            s.Swap(start, i);
            Permutations(s, start + 1, results);
            s.Swap(start, i);
        }
    }

    static void Swap(this StringBuilder s, int a, int b)
    {
        var c = s[a];
        s[a] = s[b];
        s[b] = c;
    }

    static void Main()
    {
        Console.WriteLine(String.Join(Environment.NewLine, "gerula".Permute()));
        Console.WriteLine(String.Join(Environment.NewLine, "abc".Permute()));
    }
}
